### PR TITLE
docs: unhiding HTTP/3 and adding docs

### LINKED ISSUE #12923: QUIC/HTTP3: docs
Arch overview docs, config docs, etc.

### PR DESCRIPTION
**Pull request by alyssawilk**
Unhiding HTTP/3 upstream and downstream configuration, linking to example configs, and updating docs for HTTP/3 alpha.

Risk Level: n/a
Testing: n/a
Docs Changes: yes
Release Notes: inline
#14829 #2557   #15845
Fixes #12923

Co-Authored-By: Michael Payne <michael@sooper.org>

### CODE REVIEW

**Comment by mattklein123** [comment]
Thanks in general this looks amazing! So excited to see this land, especially with example configs and maybe even a sandbox? (cc @moderation @phlax)

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (phlax) [inline api/envoy/config/listener/v3/udp_listener_config.proto:37]: i think this should be 2-space indent

File: docs/root/start/sandboxes/grpc_bridge.rst
Comment (danzh2010) [inline docs/root/start/sandboxes/grpc_bridge.rst:17]: HTTP/3 does not support the HTTP Upgrade mechanism https://tools.ietf.org/id/draft-ietf-quic-http-23.html#http-upgrade

**Review by danzh2010 — COMMENTED** [review]
Wow, that's a lot of text! Thanks for drafting it!

**Comment by phlax** [comment]
docs build is here https://storage.googleapis.com/envoy-pr/15926/docs/index.html

**Comment by alyssawilk** [comment]
oops, pushed early, /wait

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (RyanTheOptimist) [inline api/envoy/config/listener/v3/udp_listener_config.proto:41]: I think that "For example" is the start of a new sentence? If so, perhaps a period and then Capitalize "for". If it's not the start of a new sentence, perhaps it can be rephrased for clarity?

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (RyanTheOptimist) [inline api/envoy/config/listener/v3/udp_listener_config.proto:42]: nit: I feel like 'in-place' is how I am accustomed to seeing this phrase in the context of software. (As opposed to, say, "having everything in place"). Maybe? WDYT?

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (RyanTheOptimist) [inline api/envoy/config/listener/v3/udp_listener_config.proto:46]: nit: "requets" => "requests"

File: api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto
Comment (RyanTheOptimist) [inline api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:110]: For Alt-Svc, I'd be inclined to say "HTTP Alternative Services" and I might also reference the spec https://tools.ietf.org/html/rfc7838

Is it worth mention the HTTPS DNS RR here as well, since we intend to implement this? (and the spec https://datatracker.ietf.org/doc/html/draft-ietf-dnsop-svcb-https-04)

File: api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto
Comment (RyanTheOptimist) [inline api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:111]: nit: I wonder if we can improve on the phrase "If HTTP/3 is configured via HTTP/3". How bout something like:

When HTTP/3 is configured, HTTP/3 connections will be attempted. However, the connection may fail over to TCP (HTTP/1 or HTTP/2 based on ALPN) instead. If the HTTP/3 connection fails explicitly because of a protocol error, a TCP connection will be attempted instead. In addition, if the HTTP/3 connection has not succeeded after a (currently) fixed timeout, a TCP connection will be started in a parallel. The connection the succeeds first will be used.

File: api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto
Comment (RyanTheOptimist) [inline api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:118]: Just for my curiosity, what is "auto config"?

File: docs/root/configuration/best_practices/level_two.rst
Comment (RyanTheOptimist) [inline docs/root/configuration/best_practices/level_two.rst:26]: Oh wow. That's exciting! :(

File: docs/root/configuration/http/http_filters/grpc_http1_reverse_bridge_filter.rst
Comment (RyanTheOptimist) [inline docs/root/configuration/http/http_filters/grpc_http1_reverse_bridge_filter.rst:11]: Totally your call... I wonder if we should be consistent about saying either "HTTP/2 or HTTP/3" vs "HTTP/2 or above". Probably not worth doing?

File: docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst
Comment (RyanTheOptimist) [inline docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst:147]: It looks like this text is identical to the HTTP/2 text (except 3 instead of 2), but I wonder if we should remove "the"  after "utilizes"?  "Whether the cluster utilizes *http3* if configured..." WDYT?

File: docs/root/faq/load_balancing/concurrency_lb.rst
Comment (RyanTheOptimist) [inline docs/root/faq/load_balancing/concurrency_lb.rst:10]: This is not a question about this PR, but it does make me wonder.... Presumably EnvoyMobile will be a single silo? (Maybe that was in the presentation. I should go re-read those slides)

File: docs/root/intro/arch_overview/http/http3.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/http/http3.rst:27]: nit: "downstream" => "upstream"

File: docs/root/intro/arch_overview/http/http3.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/http/http3.rst:17]: nit: "via alt-svc header" => "via an alt-svc header"

File: api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto
Comment (RyanTheOptimist) [reply api/envoy/extensions/upstreams/http/v3/http_protocol_options.proto:111]: Actually, having written all that... I think that's only true when H1, H2 and H3 are all enabled. If only H3 is enabled then there's no fallback and not alt-svc checking, right? Should we mention this? (Wow this gets complex in a hurry!)

File: docs/root/intro/arch_overview/http/http_connection_management.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/http/http_connection_management.rst:23]: I think it's fine to keep this as "HTTP/2" since when Envoy was designed HTTP/3 was not yet a thing. Your call though.

File: docs/root/intro/arch_overview/upstream/connection_pooling.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/upstream/connection_pooling.rst:68]: Maybe this doesn't need to be mentioned, but TCP is only attempted after the timeout if the QUIC connection has not already succeeded, right?

File: docs/root/intro/arch_overview/upstream/connection_pooling.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/upstream/connection_pooling.rst:72]: Should we mention the HTTPS DNS RR?

**Review by RyanTheOptimist — COMMENTED** [review]
Wow, this is fantastic! Thanks for writing all this up. One or two questions but mostly some minor nits.

**Comment by alyssawilk** [comment]
I moved most of the details to the docs, where they'd been half in docs and half in APIs.
still torn about having upstream H3 details in the connection pool doc or the H3 doc - thoughts appreciated!

**Comment by alyssawilk** [comment]
cc @moderation if you have any tips on how I can make docs more clear now that you've played around, please add suggestions here!

File: docs/root/intro/arch_overview/upstream/connection_pooling.rst
Comment (RyanTheOptimist) [inline docs/root/intro/arch_overview/upstream/connection_pooling.rst:74]: nit: "or" before QUIC hints, I think

**Comment by moderation** [comment]
> Do we see a use case where folks would want to run Envoy as root without BPF? If so I'd be inclined to leave it on by default and we can add the knob iff folks come up with a use case, or get cranky about the warning that they're running it inefficiently.

Keen to ensure the use case of running without root, but with Linux capabilities set which should allow for use of BPF as per https://github.com/envoyproxy/envoy/issues/15845#issue-850909363

File: configs/envoyproxy_io_proxy_http3_downstream.template.yaml
Comment (moderation) [inline configs/envoyproxy_io_proxy_http3_downstream.template.yaml:20]: This UDP config is on the TCP listener - should it be removed?

**Comment by alyssawilk** [comment]
OK, Dan's got a fix in the works for our last blocking bug so I think this is ready for maintainer pass and can land this week.
@phlax as our docs expert would you be willing to do the !google pass?
@snowp if you'd be willing to do at least one pass as data plane owner that'd be fantastic.

**Review by alyssawilk — COMMENTED** [review]
Whoops!  Somehow I missed half of Ryan's comment on his review.  fixing now!

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (adisuissa) [inline api/envoy/config/listener/v3/udp_listener_config.proto:47]: It seems that this link, and the one above, are not rendered properly.
See example here: https://storage.googleapis.com/envoy-pr/e10ad5f/docs/api-v3/config/listener/v3/udp_listener_config.proto.html?highlight=quicprotocoloptions

**Review by adisuissa — COMMENTED** [review]
api review: small issue with the links, but other than that LGTM.

File: docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst
Comment (phlax) [inline docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst:143]: this one also i think should be double-backtick

File: docs/root/faq/configuration/timeouts.rst
Comment (phlax) [inline docs/root/faq/configuration/timeouts.rst:63]: open *a* stream window ? or the ? feels like it needs an article - maybe wrong

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (phlax) [inline api/envoy/config/listener/v3/udp_listener_config.proto:38]: im wondering if this note should be a warning - and also thinking it might be better placed in docs/root/intro/arch_overview/http/http3.rst as its linked to from various places - feels a bit arbitrary to add it to one proto

File: docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst
Comment (phlax) [inline docs/root/configuration/upstream/cluster_manager/cluster_runtime.rst:147]: and here

File: docs/root/intro/arch_overview/http/http3.rst
Comment (phlax) [inline docs/root/intro/arch_overview/http/http3.rst:2]: probably we want a page title above the upstream/downstream subtitles - something like "HTTP3 overview"

File: docs/root/intro/arch_overview/http/http3.rst
Comment (phlax) [inline docs/root/intro/arch_overview/http/http3.rst:17]: `alt-svc`

File: docs/root/intro/deployment_types/service_to_service.rst
Comment (phlax) [inline docs/root/intro/deployment_types/service_to_service.rst:17]: i think rather not emphasize the http:// address - just let it  be a link

File: tools/spelling/spelling_dictionary.txt
Comment (phlax) [inline tools/spelling/spelling_dictionary.txt:1137]: probs dont needs this ifwe make use inline literal for `alt-svc`

**Review by phlax — CHANGES_REQUESTED** [review]
@alyssawilk looks good generally

ive left a few nits inline - quite a few are backticks - some i think are probably more important than others, but feel free to ignore some

File: docs/root/intro/arch_overview/http/http3.rst
Comment (phlax) [reply docs/root/intro/arch_overview/http/http3.rst:17]: i think this is probably more important as it wants to emphasize the is the exact header

**Comment by alyssawilk** [comment]
@RyanTheOptimist / @phlax I miss anything / any other requests, or is this good to go once https://github.com/envoyproxy/envoy/pull/16320 lands?

File: api/envoy/config/listener/v3/udp_listener_config.proto
Comment (phlax) [reply api/envoy/config/listener/v3/udp_listener_config.proto:38]: im wondering if this should make use of the `experimental` annotation that has just been added

agreed that their needs to be a warning here - but i would put the fuller explanation in the ref docs and point to that - as is done in most of the the api protos - mho

File: docs/root/intro/arch_overview/http/http3.rst
Comment (phlax) [inline docs/root/intro/arch_overview/http/http3.rst:5]: whatever we keep in the api docs - i think a warning is warranted here too

File: configs/envoyproxy_io_proxy_http3_downstream.template.yaml
Comment (phlax) [inline configs/envoyproxy_io_proxy_http3_downstream.template.yaml:9]: i think that this file is not a template - ie its pure envoy config (iiuc) - iirc having template in the filename will prevent it being validated

File: docs/root/intro/arch_overview/http/http3.rst
Comment (phlax) [inline docs/root/intro/arch_overview/http/http3.rst:20]: i think the backticks should just be around `alt-svc` as that is the literal part - my rule of thumb ~is "should it be translated" if we were translating the docs

File: configs/envoyproxy_io_proxy_http3_downstream.yaml
Comment (phlax) [inline configs/envoyproxy_io_proxy_http3_downstream.yaml:17]: this needs to be a list - ie just add `-` before

Comment (alyssawilk) [reply configs/envoyproxy_io_proxy_http3_downstream.yaml:17]: annoyingly, this loads just fine in Envoy, but the precheck validator is apparently more strict than Envoy.
h/t @yuval-k who called out the same thing over on https://github.com/envoyproxy/envoy/pull/15987

Comment (phlax) [reply configs/envoyproxy_io_proxy_http3_downstream.yaml:17]: yep - it  uses the static validation in the prechecks - and yep its a bit stricter in a few ways - not sure how we can integrate/sync these two tools better - they are both useful checks

Comment (phlax) [reply configs/envoyproxy_io_proxy_http3_downstream.yaml:17]: in respect of this specific problem - i think the check is good - ie we should be explicit with arrays - apart from anything else it makes the config examples in docs much less confusing

File: docs/root/intro/arch_overview/http/http3.rst
Comment (danzh2010) [inline docs/root/intro/arch_overview/http/http3.rst:35]: Maybe also worth to mention that if BPF is not supported by the platform and multiple worker threads are configured, Envoy will route packets across workers by itself and log a warning of performance degrading, or something similar.

**Review by RyanTheOptimist — APPROVED** [review]
Hm. I could have sworn I approved several days ago, but github disagrees. LGTM

**Review by phlax — APPROVED** [review]
lgtm, thanks @alyssawilk

File: docs/root/intro/arch_overview/http/http3.rst
Comment (danzh2010) [inline docs/root/intro/arch_overview/http/http3.rst:23]: The HCM config needs to specify codec type to be `Http3` as well.

Comment (alyssawilk) [reply docs/root/intro/arch_overview/http/http3.rst:23]: Yeah, at the point you configure a listener, the rest should be auto-validated also shown by the sample config.
happy to add the detail in a follow-up but I think i'm going to land this version today :-)

**Pull request merged**
